### ISSUE TITLE
Android: Story notifications

### ISSUE THREAD

**Issue by samuelclay** [samuelclay/NewsBlur#966]
This is a big ticket feature. There are two parts to this feature, this ticket only focuses on the actual notification. Choosing notifications for feeds is the next ticket, but we can launch that later.

0) Go on the web and right-click feeds to setup "Notifications" for android. You should be able to see it as it's available to admins.

1) After `/reader/feeds?flat=true` is loaded, loop through every feed and check if 'android' is in any feed.notification_types. If so, request notification privileges on the device.

2) Once android gives you a notification token that can be used for pushing remote notifications, send that token to POST to `/notifications/android_token/` with `android_token` as the key. This is idempotent, so you can submit this over and over again with the same token. That's what we do on iOS, to handle changing tokens. This endpoint is also device-aware, so multiple devices can submit different tokens to the same endpoint and each device will receive notifications.

Use `http://www.newsblur.com/notifications/force_push?feed_id=X&count=N` to test notifications. Feel free to hit it as much as you like.

Ideally I'd like to get images in the notification preview, as I did on iOS.

![photo nov 21 5 15 53 pm](https://cloud.githubusercontent.com/assets/44229/20507281/90ab1602-b00e-11e6-8beb-84575d6943cb.png)
![photo nov 21 5 16 14 pm](https://cloud.githubusercontent.com/assets/44229/20507282/90c43402-b00e-11e6-8e2e-bc7d15753183.png)
![photo nov 21 5 16 17 pm](https://cloud.githubusercontent.com/assets/44229/20507283/90c4994c-b00e-11e6-979a-9742c65db11a.png)

**Comment by samuelclay** [comment]
Mark read and Save story actions do not open the app, they just silently work. Tapping on the notification opens the feed and then finds the story.

**Comment by dosiecki** [comment]
Is there any background documentation on what a token is and the design rationale for this?  The app already has all the metadata needed to know when to post a notification that a feed has a new story without the whole extra round-trip to the server.  That is, if `notification_types` has the android flag, can't the app just post a notification when those feeds have a new story?  

Also, from a system standpoint, one-notification-per-story will likely not work.  Individual notifications can be heavyweight to process on some devices and the system will limit them and/or break if they are overused.  The typical design pattern on Android for a UX like this is to have a single notification that says "you have N new stories" and it changes as the count does rather than stacking up notifications for each new story.

**Comment by samuelclay** [comment]
The server has to push the notification to Google's servers. I need that token to send. I looked this up, seems like the GCM is now Firebase. 

And one notification per story is exactly what I'm going for here. People can choose which feeds are getting notifications. And they can limit them to Focus Stories only. This is not for all stories, just a few that a user is highlighting.

**Comment by samuelclay** [comment]
Here's the token I was referring to: https://firebase.google.com/docs/cloud-messaging/android/first-message

**Comment by samuelclay** [comment]
And if you take a look at the python library for FCM, it just takes a token and the message, exactly like I do for iOS: https://pypi.python.org/pypi/pyfcm/

**Comment by dosiecki** [comment]
From a design standpoint, using GCM is very overkill.  The app already gets all the necessary data when it does normal background syncs.  GCM is more for helping build apps that don't already have a working BG sync API.  We can issue notifications just fine with our existing sync loop, and adding GCM support would take the complexity of the app to a whole new level.

Also, bear in mind that not all Android devices support GCM, but the do all support basic notfications that would be built on your existing API infrastructure.

**Comment by samuelclay** [comment]
I don't want to use the bg sync, since that's not nearly as fast as GCM/FCM. I looked at how complicated it is to integrate and it seems like it'll take less than a day, same as it did on iOS. I'm not sure I see where the complexity is. NewsBlur servers send a FCM message with the story_hash, feed_id, title, and ~1.5k of content. The device displays those stories and when the user selects one, the app opens up, opens the feed, finds the story, and that's that.

**Comment by dosiecki** [comment]
GCM does slightly speed up notifications by a few seconds/minutes compared to our existing sync framework, but in a majority of cases it will be perceptually similar.  One tangible benefit of GCM is that it hides a fraction of our app's battery and network usage from the user.

The downsides are pretty rough, though:

On Android, you don't actually talk to the GCM API yourself; you are required to delegate syncs to a whole extra framework on the device and use a hybrid userspace/kernel library to communicate with it.  Built-from-source Android or non-GOOG-branded Android (read: most of Asia) don't support it. (none of my three daily-driver devices support GCM, though all 'stock' US/EU Nexus/Pixel devices should, so long as they are current on system updates)  Use of GCM would also make the app ineligible for distribution via F-Droid unless we create and maintain a GCM-free branch.

**Comment by samuelclay** [comment]
Ok, build it by scanning for 'android' in `notification_types` and then do a double-check for `is_focus` and only notify focus stories then. I'll see how that works in practice, since I have the stories coming in on iOS and web already. I'm not convinced that we shouldn't implement GCM, but I'd like to try this first.

**Comment by samuelclay** [comment]
If you like I have the last notification_date, although for your purposes it will always be wrong because by the time you get it, the notifications will have been sent and the date too far forward.

**Comment by dosiecki** [comment]
Pull #970 implements the notification lifecycle as discussed.  It is currently gated behind having the Download Stories option checked, since normal users don't have an enablement option yet.

Next we will have to discuss the actual style and content of the notifications.  The UI depicted in the iOS screenshots is, in a number of ways, incompatible with the Android platform that the vast majority of users are running.  We will have to design the UI around the platform limitations on notification grouping, amount of content that can be presented, etc.  As a placeholder, the notification shows the number of new stories and what feeds they are from.  Tapping it takes you into the app.  Let's discuss specific additional UI features one by one to figure out which can be done for which users.

**Comment by dosiecki** [comment]
One of the first issues to tackle is notification cardinality.

While it is easy to implement the posting of a separate notification for every new story, this is recommended against.  Multiple notifications will be grouped by the OS, showing a single, summary notification that often only shows the app name, which isn't a terribly clear UX.  Thus, the platform guidelines suggest internally handling a situation where we have more than one notification to show and creating and issuing a single summary notification that can still have working title / body / actions.  This grouped style is the boilerplate I used in the beta impl.  What needs to be explored is whether there is any utility in also having a specialized, more detailed notification style for single stories, or if the case of a single, new story will be rare compared to having n>=2 new stories at any given time.

It is also worth noting that Android 7 introduced a whole new framework for notifications that does allow for significantly better handling of high cardinality.  It is well suited for apps that post lots of small, individually consumable things like messages, or stories.  However, at present, about 0.3% of users are on Android 7+, and it will probably be at least a year before even a majority of our advanced users would see this impl.  Still worth looking ahead, to, though.

**Comment by samuelclay** [comment]
I've had the notifications build loaded all night but I'm not receiving many notifications. I got one that said "1 New Story  - Daring Fireball" but my iphone, which is configured with the same notifications, has received tons.

We should be splitting up each story into a single notification. This notification feature is not for general interest feeds, this is about high importance feeds that deserve immediate attention. This is also why I am starting from a position of not wanting the 10-minute background fetch. Turns out my phone had offline stories disabled. This should be auto-enabled for every phone. So if I had setup Android notifications, they wouldn't have worked. It's possible I turned it off, but I can't imagine why I would do that. People who are worried about their storage (and on iOS, a full 10,000 stories takes 1-2GB of space) can turn it off.

Notifications need a story title and a small excerpt of the story content. I've seen other notifications on Android that work this way, namely Messenger, the NYTimes, and Twitter. I don't mind if it only hits half of our users. It's incentive to upgrade their phones. I'm trying to make the experience for people on tech that supports the latest [2-3 versions] of OS features, even if that's only 50% of users. The rest can stick with an older, unmaintained version.

![2016-12-02 19 22 39](https://cloud.githubusercontent.com/assets/44229/20847179/caaf34a2-b881-11e6-85d7-ece794ef58c1.png)

Like the "Screenshot captured" notification, I'd love to see the story image as the larger image and the feed favicon as the smaller icon. When no story image, use the favicon large. And the title can be a single line with both feed_title and story_title: "Publisher: Story Title"

**Comment by dosiecki** [comment]
> I'm not receiving many notifications
> Turns out my phone had offline stories disabled

Please see my release notes.  I temporarily gated notifications behind that option so you have a way to disable them without the public UI containing spoilers for the new feature.

> This should be auto-enabled for every phone

Will do once the option is in place.

> We should be splitting up each story into a single notification

As mentioned above, this has serious negative effects for the app when implemented for pre-Android 7 devices (currently 99.7% worldwide).  Showing a single notification for a single story is fine.  Showing one for each of 2 or 3 will be shown on some devices (mostly tablets) without problem.  More than that will get rolled up and the content hidden.  Significantly more than that may cause the system to stop showing your notifications.  As mentioned, the solution is to bundle up more than 1 (or 2, if you want to risk it) into a summary notification so you can at least stll get nontrivial data into the status bar.  This lets notifications keep going through with rich content over which we have control.

> Notifications need a story title and a small excerpt of the story content

Yep, working out how to do this in the best-supported way once we tackle the cardinality issue.

> Like the "Screenshot captured" notification, I'd love to see the story image as the larger image and the feed favicon as the smaller icon. When no story image, use the favicon large

Doable with some extra custom view code work for a subset of devices.  I will follow up with you with some time estimates for how much various degrees of support will take.

**Comment by samuelclay** [comment]
I'm seeing 4 NYTimes notifications in the same day. I realize they might be rolled up, I'm OK with that. I do not want to truncate the notifications, that's not the point of them. This is for immediate action stories.

"1 New Story" should be the "Feed Title: Story Title"
"Daring Fireball" should be the first 200 characters of the story.
The logo should be the `story_image` or the favicon, with NewsBlur app icon logo in lower right.

![2017-01-03 23 02 08](https://cloud.githubusercontent.com/assets/44229/21628302/50027c9e-d1d3-11e6-8f5b-a9784626dcb0.png)

**Comment by dosiecki** [comment]
Took me a few days to get my hands on a genuine 7.1 device.  I just rolled back the smoke test code from mainline to support the bugfix release and to clean the slate to work on pretty much all of the things you mention.

My next goal is to get a second smoke test build with all of the above to show some examples of how high-cardinality notifications look on various platforms and to explore the difference in limits.

**Comment by samuelclay** [comment]
I've noticed that the NYT only shows the latest 4 or 5 stories. Can we retract story notifications so that we keep the number down to most recent 5 at all times?

**Comment by dosiecki** [comment]
Update on this: reworking the favico fetching system a bit so they can be used in notifications.  The existing fetcher was closely tied to handing them off to a View, but the notification system needs them as an in-memory Bitmap.  Cleaning things up so that the icons can be efficiently fetched and used by both subsystems.

**Comment by samuelclay** [comment]
Ok, I'm pushing out the current build as v5.0. Since notifications are not yet launched, we don't need to do anything about them then, since they aren't shown anywhere unless you have enabled notifications on the web.

**Comment by dosiecki** [comment]
Next PR contains the second alpha of notification support.  Note that it is still gated behind story sync being enabled so beta users don't get the confusing notification option when they have no way of actually turning them on from web.  I will expose the proper option when you are ready to launch.

> I've noticed that the NYT only shows the latest 4 or 5 stories. Can we retract story notifications so that we keep the number down to most recent 5 at all times?

I did a ton of research into how many notifications an app can push on various platform versions before it encounters problems.  The behaviour, as you might expect, varies a ton from device to device.  In brief, though, so long as notification throughput doesn't exceed about 50 per wake cycle, we won't get banned.  Devices start rolling up notifications when more than about 6 are displayed (or how many will fit on screen), with varying visual effects.  I think the NYT folks have it spot on by limiting to 5, since that is about as many as can politely fit on screen for a high-percentile device size.  I've implemented the same limit of 5, giving priority to focus stories and newer stories over older ones.

> The logo should be the story_image or the favicon, with NewsBlur app icon logo in lower right.

Favicons are now in place.  What background colour would you like for the overlay/default icon? (In your example, NYT was using black and Twitter was using a blue.  Android limits you to a solid colour with a while foreground.  The unset default varies by device and is usually a grey or off-blue.)

So far, all of this has been implemented for the 99% case of pre-Android 7.  We should definitely enhance the app to use v7+ features, though, because they greatly, greatly improve the UX for apps that push more than one notification at a time.  (on a personal note, I find the pre-7 Android UX for apps that push multiple notifications to be so ugly, I turn them off entirely)

**Comment by samuelclay** [comment]
This is awesome, I'm excited to play with it.

**Comment by dosiecki** [comment]
deployed to Prod
